Enable ServerApp to discover ExtensionApps (and their config).

Zsailer opened the pull request:
## Summary 

This PR enables the ServerApp to discover ExtensionApp's before the server's Tornado Web Application is configured. 

Other smaller changes:
1. Extensions' `default_url` changed to `extension_url`.
2. The order of paths for searching enabled extensions was reversed from user > env > sys-prefix to sys-prefix > env > user. This was a lingering fix in accordance with [JEP #29](https://github.com/jupyter/enhancement-proposals/blob/master/jupyter-server/jupyter-server.md) 
3. Extensions' loading function, `load_jupyter_server_extension`, changed to `_load_jupyter_server_extension` (a private function prefixed with `_`).
4.  ServerApp now uses extensions' `_jupyter_server_extension_paths` function to discover server extensions. Previously, this was only used to toggle extensions. 
5. Added an optional `"app"` key to the metadata returned by `_jupyter_server_extension_paths`.
6. ServerApp now initializes extensions first in it's `initialize` method. Then, constructs a WebApplication.

## tl;dr Why?
 
There's a challenge with the order in which the Server and Extensions are loaded. Previously, the order is:
1. Server is initialized
2. Server config is loaded
3. Tornado WebApplication is created
4. Extensions are initialized through `load_jupyter_server_extension`
5. Extension config is loaded

The issue is that, during a transition to jupyter server, extension apps that try configuring serverapp do not load their config early enough (step 5). The serverapp will already have created it's WebApplication and HTTPServer objects before the extension can change their attributes. E.g. if you want to change the `port` of the server through a CLI like `jupyter lab --LabApp.port=8890`, you cannot currently do this. The port value will be loaded too late. The server will run on 8888, and the trait will change after this server starts.

This PR adds a step after step 2, where the ServerApp can *discover* ExtensionApp subclasses listed in jpserver_extensions. The ServerApp can then call its `load_config_file()` method and discover server-specific config from the ExtensionApp.

1. Server is initialized
2. Server config is loaded
3. **Extension are found+stored in new attribute and ExtensionApp are initialized (thus, config loaded).**
4. Tornado WebApplication is created
5. Extensions are loaded through `_load_jupyter_server_extension` function.

The added benefit is that it allows me to intercept the ServerApp config and shim traits using the shim layer added in Zsailer/nbclassic#7. 

## How

This adds an optional key:value pair to the extension metadata dict, `app: <ExtensionApp subclass`. If an ExtensionApp is listed in `jpserver_extension`, this key will be found in the extension metadata and discoverable by ServerApp. Then, the ServerApp's initialize method can pick up these extensions and load their config before creating a webapp.

*My apologies for the auto-formatting diffs! 😦 If the diff is too difficult to look at, I can highlight the chunks that matter*

Zsailer commented on jupyter_server/extension/application.py:
```diff
@@ -175,13 +175,20 @@ def _validate_extension_name(self, value):
+            "module": cls.__module__.split('.')[0],
+            "app": cls
+        }]
+
```
New method that builds extension metadata.

Zsailer replied:
I have a few thoughts on this, but I think it's worth discussing in tomorrow's meeting. After that discussion, I'll follow up here with notes.

Zsailer replied:
I think it makes the most sense to explicitely have extensions authors write this function, so I've removed this from ExtensionApps.

Zsailer commented on jupyter_server/serverapp.py:
```diff
@@ -1491,12 +1491,32 @@ def init_server_extension_config(self):
+                app.load_config_file()
+                # Pass any relevant config to the
+                # serverapp's (parent) config.
+                self.update_config(app.config)
```
Logic to load ExtensionApp config.

Zsailer commented:
@vidartf and @echarles, I labeled you as reviewers since you've been involved most with the discussion... hopefully you don't hate me too much 😉

echarles commented:
Hi Zach, I have done a few tests with this branch.

My jlab work is completed since yesterday and works fine with master. With this branch, I face 2 issues

[1] Running jupyter lab does not return exception, but a 404 in the browser.
[2] The examples (in jlab examples/ folder) do not start anymore and return in `_get_server_extension_metadata   raise KeyError(u'The Python module {} does not include any valid server extensions'.format(module))`

I have then used the example branch and the extensions examples work fine with this branch.

I would suggest:

[1] merge https://github.com/jupyter/jupyter_server/pull/117 so it is easier to have a debuggable baseline upon the tests.
[2] gimme a few days to debug what is going on with the jlab port.

echarles commented:
@Zsailer This is the exception I receive when I run an example. Would it be related to the new method introduced here for the metadata or the _jupyter_server_extension_paths? If such, any tip to solve it?

```python
Traceback (most recent call last):
  File "main.py", line 37, in <module>
    ExampleApp.launch_instance()
  File "/Users/echar4/datalayer/repos/jupyter-server/jupyter_server/extension/application.py", line 450, in launch_instance
    load_other_extensions=cls.load_other_extensions
  File "/Users/echar4/datalayer/repos/jupyter-server/jupyter_server/extension/application.py", line 366, in initialize_server
    serverapp.initialize(argv=argv, load_extensions=load_other_extensions)
  File "<decorator-gen-7>", line 2, in initialize
  File "/opt/datalayer/opt/miniconda3/envs/jlab-server/lib/python3.7/site-packages/traitlets/config/application.py", line 87, in catch_config_error
    return method(app, *args, **kwargs)
  File "/Users/echar4/datalayer/repos/jupyter-server/jupyter_server/serverapp.py", line 1684, in initialize
    self.init_server_extension_config()
  File "/Users/echar4/datalayer/repos/jupyter-server/jupyter_server/serverapp.py", line 1492, in init_server_extension_config
    _, metadata = _get_server_extension_metadata(modulename)
  File "/Users/echar4/datalayer/repos/jupyter-server/jupyter_server/extension/serverextension.py", line 381, in _get_server_extension_metadata
    raise KeyError(u'The Python module {} does not include any valid server extensions'.format(module))
KeyError: 'The Python module __main__ does not include any valid server extensions'

echarles commented:
To reproduce, you need those 2 branches https://github.com/jupyterlab/jupyterlab/pull/7416) and https://github.com/jupyterlab/jupyterlab_server/pull/79.

Then go to `jupyterlab/examples/app` and run `python main.py`

echarles commented:
I have added this in the jupyterlab_server handler and th example work back. Still, I receive a 404 with the full jupyter lab command even adding _jupyter_server_extension_paths on the lab handler. The good think is that `jupyter lab -h` works now, it was failing before.

```
    @staticmethod
    def _jupyter_server_extension_paths():
        return [{
            'module': 'jupyterlab_server'
        }]
```

echarles commented:
All good now - Lab is working again  @@. I have added a TODO to discuss later on the impact
https://github.com/datalayer-contrib/jupyterlab/commit/86b89ec7c54190a74efa880ceb9cd913633cda14#diff-f7b3ec31a3cdaf9ff4b443cf6ff0d38eR563

vidartf commented on jupyter_server/extension/application.py:
```diff
@@ -354,20 +352,23 @@ def _prepare_templates(self):
         # Get a jupyter server instance
         serverapp = ServerApp.instance(**kwargs)
+        # Add extension to jpserver_extensions trait in server.
+        mod = cls._jupyter_server_extension_paths()[0]['module']
```
Is the variable meant to be used?

Zsailer replied:
It *was* being used as the key to the dictionary in the next line, but I've changed this logic a bit so this no longer applies. Resolving.

vidartf commented on jupyter_server/extension/application.py:
```diff
@@ -377,9 +378,12 @@ def initialize(self, serverapp, argv=[]):
         self.serverapp = serverapp
-        
-        # Initialize the extension application
-        super(ExtensionApp, self).initialize(argv=argv)
```
Will this be significant for mixin MROs? I.e. if a mixin is a sibling ancestor as `ExtensionApp`, will this make the order the siblings are declared (left-right / right-left) be dictated? Is it already dictated? If so, is it documented?

Zsailer replied:
Good question... I need to think about this some more.

vidartf replied:
Did you figure something out? Punting is acceptable if deemed unimportant :)

Zsailer replied:
This PR significantly simplified how ExtensionApps are initalized and loaded. There is no need to call `super(...).initialize` anymore, since ExtensionApp's defer CLI parsing and config loading to the Server as it's parent app.

vidartf commented on jupyter_server/extension/application.py:
```diff
@@ -411,18 +415,18 @@ def stop(self):
-        extension = cls()
-        extension.initialize(serverapp, argv=argv)
+        extension = cls(parent=serverapp)
+        extension.initialize(serverapp)
```
This seems pretty significant, what is the logic behind it? I.e. why are we no longer passing `argv` to the extension?

Zsailer replied:
Argv is parsed by the ServerApp directly and the ExtensionApp traits are inherited by this ServerApp via `parent=serverapp`. Previously, we were double parsing the args—once in the ServerApp and again in the ExtensionApp.

Zsailer commented:
Checkout #188. There are some significant inconsistencies in our extension loading framework. I'd like to revisit them, fix it here, and rewrite our documention in this PR. Adding a [WIP] tag until I've finished.

Zsailer commented:
@echarles Thanks a bunch for testing out this PR. I've been making some significant changes to the extension loading API as it has many inconsistencies. I'm hoping we smooth some of these issues.

Zsailer commented:
I think this is ready for final review.

vidartf commented on jupyter_server/extension/application.py:
```diff
@@ -181,8 +181,8 @@ def _validate_extension_name(self, proposal):
-    aliases = aliases
-    flags = flags
+    # aliases = aliases
+    # flags = flags
```
Whenever I see commented out code like this, I assume this was some temporary measure? Could you either add a comment explaining why this code is still here, or remove it?

Zsailer replied:
Aliases and flags are now empty by default in ExtensionApps

vidartf commented on jupyter_server/extension/application.py:
```diff
@@ -284,8 +274,10 @@ def initialize_templates(self):
-        the object to the webapp's settings as `<extension_name>_config`.  
+        the object to the webapp's settings as `<extension_name>_config`.
         """
+        # Verify all traits are up-to-date with config
```
I wish this comment was more about *why* instead of *what*. The comment also seems somewhat at odds with the method name (in which direction does the information flow, from traits -> config, or the other way around?).

Zsailer replied:
The lines above were removed. they weren't need after all.

vidartf commented on jupyter_server/extension/handler.py:
```diff
@@ -12,19 +12,29 @@ def get_template(self, name):
 
+    @property
+    def extensionapp(self):
+        key = "{extension_name}".format(extension_name=self.extension_name)
```
Does this do anything other than copy the string?

Zsailer replied:
Yuck! 😆 That was bad on my part. haha Fixed!

vidartf commented on jupyter_server/extension/handler.py:
```diff
@@ -12,19 +12,29 @@ def get_template(self, name):
+    @property
+    def serverapp(self):
+        key = "serverapp"
+        return self.settings[key]
```
I'm confused by the property names used here: `serverapp` is a property that maps onto `settings`. One implies a ServerApp instance, the other a settings value (not typically run-time objects?).

Zsailer replied:
Hm, the weird thing is that we pass all kinds of run-time objects to the webapp settings (kernel manager, contents manager, session manager, etc.) I agree, this probably isn't best practice but would take an extension refactor.

vidartf replied:
So wait, what is `type(self.settings[key])` here? `ServerApp` ?

vidartf commented on jupyter_server/serverapp.py:
```diff
@@ -1487,31 +1490,96 @@ def init_server_extension_config(self):
-                    if self.reraise_server_extension_failures:
-                        raise
-                    self.log.warning(_("Error loading server extension %s"), modulename,
-                                  exc_info=True)
```
Should we not keep this exception guard? Or is its implementation moved? I.e. shouldn't the actual calling of the `load_jupyter_server_extension` function be guarded?

Zsailer replied:
Depends on what you mean. It *is* guarded in that, if it no load_jupyter_server_extension function is found, the serverapp doesn't crash but throws (more helpful) console log warnings. This PR improves the overall messaging around server extension failures.

vidartf replied:
I meant: Previously, this code had a `try: load_jupyter_server_extension(...) except Exception`. Now it has:
```python
            try:
                # This function was prefixed with an underscore in in v1.0
                # because this shouldn't be a public API for most extensions.
                func = getattr(extension, '_load_jupyter_server_extension')
                func(self)
                self.log.debug(log_msg)
            except AttributeError:
                # For backwards compatibility, we will still look for non
                # underscored loading functions.
                func = getattr(extension, 'load_jupyter_server_extension')
                func(self)
                self.log.debug(log_msg)
            except:
                warn_msg = _(
                    "{extkey} is enabled but no "
                    "`_load_jupyter_server_extension` function "
                    "was found.".format(extkey=extkey)
                )
                self.log.warning(warn_msg)
```

So if the `load_jupyter_server_extension` throws an error (`AttributeError` or not), the error message will be less informative.

vidartf commented on tests/extension/test_serverextension.py:
```diff
@@ -91,34 +91,36 @@ def test_merge_config(
-def ordered_server_extensions():
-    mockextension1 = SimpleNamespace()
-    mockextension2 = SimpleNamespace()
+# ### NEED TO REFACTOR
```
Is this intended to go in as-is?

vidartf commented on jupyter_server/serverapp.py:
```diff
@@ -1487,31 +1490,96 @@ def init_server_extension_config(self):
+                            self.log.warn(log_msg)
+                except KeyError:
+                    # A KeyError suggests that the module does not have a
+                    # _jupyter_server_extension-path.
```
Would it be cleaner to put this somewhere closer to the actual lines that raise this exception to ensure we don't misattribute other `KeyError`s?

Zsailer replied:
:+1: refactored slightly to fix this. Thanks!

vidartf commented on jupyter_server/serverapp.py:
```diff
@@ -1487,31 +1490,96 @@ def init_server_extension_config(self):
+                        extapp = metadata.get('app', None)
+                        extloc = metadata.get('module', None)
+                        # Load the extension module.
+                        if extapp and extloc:
```
I read this logic as "if an ExtensionApp, use new initialization path, if not, use old one". It is unclear to me if we should initialize *all* `ExtensionApp`s this way, or only the one being initialized as the main app?

Zsailer replied:
I think all `ExtensionApp`s should be initialized the same way by the server... currently that's by creating and instance of the app and calling `link_to_serverapp`. 

I can see that it's a bit weird that I'm checking `extloc` here too. My intention was that `extloc` should always be found in `metadata`. If it's not found, the extension should not be initialized and a warning thrown below.

echarles commented:
@Zsailer With a6096ed, the `default_url` does not work anymore (was working before). Is this a reegression?

Zsailer commented:
@echarles, yes I dropped `default_url` from the `ExtensionApp`. This was intended—and I think it makes the most sense, actually. The `default_url` should remain unique to the `ServerApp`. The extension will have its own landing URL, and it should be distinguished from the `ServerApp`. Sometimes, these will be the equal (e.g. when `ExtensionApp.launch_instance` is called), but not always.

Previously, I was doing some terrible hacks to make sure the ServerApp's default_url and Extension's default_url were equal and that the Server's `browser_open_file` opened to the correct landing page. Now, I realize it makes sense to use **configure** the server's default_url to match the *extension_url* when `Extension.launch_instance` is called, and ServerApp will handle all this setup automatically+appropriately. In the end, this drastically simplified the `ExtensionApp`.

I will update examples + tests in the ServerApp repo. I'll also need to fix nbclassic. You'll need to update this trait in Lab.

Zsailer commented:
@echarles Also, I fixed the bug/issue we were having [here](https://github.com/Zsailer/jupyter_server/pull/6#issuecomment-602069694) calling `initialize` twice! Thanks for calling that out!

echarles commented:
> @echarles, yes I dropped default_url from the ExtensionApp...

@Zsailer Sure, i can understand the reasoning. Now, when I run `jupyter lab` the default server page is shown, this is not what we want. How can I ensure that the lab page is shown by default when I run `jupyter lab`?

echarles commented:
@Zsailer Would it be `extension_url` trait on ExtensionApp? If one extension is enable, would the server ensure that is opens the page at that URL by default?

Zsailer commented:
@echarles and @vidartf, this is ready for another (hopefully final) round of review. 

I've updated many tests and examples too. Specifically for you @echarles, I've made sure that all entrypoints to the examples work as you'd expect. Let me know if you have questions.

echarles commented:
Hi @Zsailer, just tried your latest commit (pip installed -e the jupyter_server and the examples) and it unblocked me for the work on jupyterlab as the launch via modules works great. Thx a bunch for that!

A few quick comments for fix on the current branch status:

- Running `jupyter server` returns ` .../serverapp.py", line 1553, in init_server_extensions for metadata in metadata_list: UnboundLocalError: local variable 'metadata_list' referenced before assignment`
- When I launch `jupyter simple-ext1 --SimpleApp1.configA="ConfigA from command line"`, the computed configA is the one from the config file, not the one given from the CLI which should take precendence.
- `jupyter simple-ext2` fails to start - I think `extension_url` has not been defined on SimpleApp2

Good news: `jupyter simple-ext11 --hello --simple11-dir any_folder` works as expected :)

echarles commented:
@Zsailer I also see that the py fixture constructs has changed and that I would need to update what I am doing in jupyterlab. I have given a try but instead of persisting, a better strategy (for me :) ) would be to let you update the https://github.com/Zsailer/nbclassic/pull/7 with the needed changes. It will then be easier for me to identify the impact and replicate in jlab. WDYT?

Zsailer commented:
@echarles Thanks for testing this out! I'm working on the issues you mentioned above. 

> let you update the Zsailer/nbclassic#7 with the needed changes. It will then be easier for me to identify the impact and replicate in jlab

Definitely! :) I'll let you know once I get this fixed.

Zsailer commented:
@echarles, I fixed the items above. 

Checkout Zsailer/nbclassic#8 to see how I adjusted nbclassic's tests. In short, ExtensionApp's no longer act like a normal entrypoint. Instead, they redirect the entrypoint to ServerApp and retrieve their argument config from ServerApp's loaded config objects. If you want to test the ExtensionApp's CLI, pass those args to the ServerApp's `argv` pytest fixture with the correct prefix.

echarles commented:
Thx @Zsailer. Quickly tried the example with your last commit and I receive `'SimpleApp1' has no attribute 'load_jupyter_server_extension'`. Now looking at the tests fixture changes

Zsailer commented:
@echarles , oops—looks like I missed some leftover code on the examples. Should be fixed. 

I think it might be work adding tests that run the examples directory at some point (in a future PR). That way, CI can pick up these issues rather than having us run these manually.

afshin commented on examples/simple/simple_ext1/application.py:
```diff
@@ -1,19 +1,19 @@
     # The name of the extension.
     extension_name = "simple_ext1"
 
     # Te url that your extension will serve its homepage.
```
`Te url` ... pre-existing typo

Zsailer replied:
```suggestion
    # The url that your extension will serve its homepage.
```

Zsailer commented:
The failing test is unrelated to this PR. This test has been flaky since merging #201 and #191.

This PR has approved and open for a little while, so I'm going merge. Thanks all for the review!

kevin-bates commented:
I fired off a restart on this, just to see if py3.7 terminal test succeeds. I suspect the timeout on that particular test may need to be increased.  It's too bad there isn't a way to restart a single job with GH actions - at least that I could find.

vidartf commented on jupyter_server/serverapp.py:
```diff
@@ -1494,31 +1521,109 @@ def init_server_extension_config(self):
+                # This function was prefixed with an underscore in in v1.0
+                # because this shouldn't be a public API for most extensions.
+                func = getattr(extension, '_load_jupyter_server_extension')
+                func(self)
```
This is the line I meant should be guarded with an `"Error loading server extension"` warning.

vidartf commented on jupyter_server/serverapp.py:
```diff
@@ -1494,31 +1521,109 @@ def init_server_extension_config(self):
+                func = getattr(extension, 'load_jupyter_server_extension')
+                func(self)
+                self.log.debug(log_msg)
+            except:
```
And this one will swallow Ctrl+C kills.
